File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/TrackDirection.java
package inc.reactor.sdk;

import java.util.Locale;
import java.util.Optional;

/**
 * Which way media flows on a track.
 *
 * <p>From the client's point of view: {@link #RECVONLY} is the model producing and this client
 * receiving, {@link #SENDONLY} is this client producing and the model receiving.
 */
public enum TrackDirection {
    /** This client sends. Push frames into it, after publishing. */
    SENDONLY,
    /** This client receives. Register a frame handler on it. */
    RECVONLY;

    /**
     * @param direction {@code "sendonly"} or {@code "recvonly"}, as the session declares it
     * @return the matching constant, or empty for a direction this SDK does not know
     */
    public static Optional<TrackDirection> of(String direction) {
        try {
            return Optional.of(valueOf(direction.toUpperCase(Locale.ROOT)));
        } catch (IllegalArgumentException unknown) {
            return Optional.empty();
        }
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/TrackDirection.kt
package inc.reactor.sdk.android

/**
 * Which way a track flows, from the application's point of view.
 *
 * [SENDONLY] is a slot the app pushes into; [RECVONLY] is one the model generates into. The model
 * declares both, and the direction is what makes `pushFrame` on a recvonly track a mistake this
 * SDK refuses rather than a push the native layer silently drops.
 */
public enum class TrackDirection(
    internal val wire: String,
) {
    /** The application sends; the model receives. */
    SENDONLY("sendonly"),

    /** The model sends; the application receives. */
    RECVONLY("recvonly"),
    ;

    internal companion object {
        fun of(wire: String?): TrackDirection? = entries.firstOrNull { it.wire == wire }
    }
}
